events/renderers/docx.py

```diff
@@ -0,0 +1,269 @@
+        dates[date_range].append(event)
+
+    dates = OrderedDict(sorted(dates.items()))
+
+    return dates
+
+
+class DOCXRenderer(renderers.BaseRenderer):
+    media_type = (
+        'application/vnd.openxmlformats-officedocument'
+        '.wordprocessingml.document'
+    )
+    format = 'docx'
+    charset = None
+    render_style = 'binary'
+
+    def render(self, data, media_type=None, renderer_context=None):
+        document = Document()
+        parsed_events = []
+
+        query_params = renderer_context['request'].query_params
+
+        event_parser = EventParser()
+
+        if len(data) < 1:
+            raise APIException('No results.')
+
+        first_location = data[0]['location']
+        for raw_event in data:
+            parsed_events.append(event_parser.parse_event(raw_event))
+
+            if raw_event['location'] != first_location:
+                raise APIException('Only one location allowed.')
+
+        # This is here to allow for this to be expanded to include multiple
+        # locations in the future.
+        locations = group_by_location(parsed_events)
+
+        for location in locations:
+            locations[location] = group_by_date(locations[location])
+
+        # We need to get the daterange for the entire document, which is
+        # determined by either the query or the actual events.
+        try:
+            query_start_date = dateutil.parser.parse(query_params.get('start'))
+        except:
+            query_start_date = event_parser.earliest_date
+
+        try:
+            query_end_date = dateutil.parser.parse(query_params.get('end'))
+        except:
+            query_end_date = event_parser.latest_date
+
+        total_date_range = DateRange(query_start_date, query_end_date)
+
+        for location, dateranges in locations.items():
+            document.add_heading(str(location), 0)
+            document.add_paragraph(str(total_date_range))
+
+            for daterange, events in dateranges.items():
+                document.add_heading(str(daterange), 1)
+
+                for event in events:
+
+                    # This is here to prevent 00:00-00:00 from being shown.
+                    if event['start_time'].hour == event['end_time'].hour == 0:
+                        document.add_heading(event['name'], 2)
+                    else:
+                        document.add_heading(
+                            '%s-%s %s' % (
+                                event['start_time'].strftime('%H:%M'),
+                                event['end_time'].strftime('%H:%M'),
+                                event['name'],
+                            ),
+                            2,
+                        )
+
+                    document.add_paragraph(event['description'])
+                    if event['price']:
+                        document.add_paragraph(event['price'])
+
+        filename = '%s-%s-%s.docx' % (
+            slugify(event['location'].name),
```

Comment:
It's not good form to use a variable introduced by a `for` block outside it. (`event` in this case)